<h1 style="margin: auto; font-weight: bold; padding: 30px 30px 0px 30px; color:#000;" align="center">SCHOOL DROPOUT IN SOGAMOSO</h1>
<h3 style="width: 100%; text-align: center; float: left; font-size: 24px; color:#000;" align="center">| EDA Notebook |</h3>
<h4 style="width: 100%; text-align: center; float: left; font-size: 24px; color:#000;" align="center">Correlation One - DS4A</h4>
<p style="text-align:center;">In this notebook will be conducted an EDA in order to understand the data and how those data is related with the Schoold Dropout in Sogamoso, Colombia.</p>

## 1. Libraries

In [1]:
# General
import pandas as pd
import numpy as np

## Own specific functions 
from functions import *

## 2. Importing files

In [2]:
df_students_2019 = pd.read_csv("datasets/alumnos_s_oficial2019.csv", encoding="cp1252", sep=";")
df_students_2020 = pd.read_csv("datasets/alumnos_s_oficial2020.csv", encoding="cp1252", sep=";", low_memory=False)
df_students_2021 = pd.read_csv("datasets/alumnos_s_oficial2021.csv", encoding="cp1252", sep=";")
df_students_2022 = pd.read_csv("datasets/alumnos_s_oficial2022.csv", encoding="cp1252", sep=";")

### 2.1 Initial analysis

<div class="alert alert-block alert-info">
    <p>Let's make a initial analysis over the datasets</p>
</div>

In [10]:
df_analysis(df_students_2019, key_columns=["NUI"], analysis_type="complete")


Analysis header
--------------------------------------------------------------------------------
- Dataset shape:			 21710 rows and 31 columns
- Total of NaN values:			 98130
- Percentage of NaN:			 14.58 %
- Total of infinite values:		 0
- Total of empty columns:		 1
- Total of empty rows:			 0
- Total of full duplicates rows:	 0
- Unique indexes:			 True
- Memory usage:				 5.1+ MB

- The key(s):	 ['NUI'] is not present multiple times in the dataframe. 
		 It CAN be used as a primary key.


Detailed analysis
--------------------------------------------------------------------------------


name     type  records  unique  # NaN     % NaN  \
0                        ANO    int64    21710       1      0   0.00000   
2                     ESTADO   object    21710      10      0   0.00000   
3                  JERARQUIA   object    21710       3      0   0.00000   
7                     SECTOR   object    21710       1      0   0.00000   
22                    GENERO   object    21710       2      0   0.00000   
19                       NUI   object    21710   21710      0   0.00000   
17                  FECHAINI   object    21710   16656      0   0.00000   
1                        ETC   object    21710       1      0   0.00000   
20                   ESTRATO   object    21708       8      2   0.01000   
29              DISCAPACIDAD   object    21705      15      5   0.02000   
24           FUENTE_RECURSOS   object    21705       5      5   0.02000   
23       MATRICULACONTRATADA   object    21705       6      5   0.02000   
13                 GRADO_COD  float64    21701      19      9   0.04000   
15                    MODELO   object    21695       7     15   0.07000   
8                       SEDE   object    21695      57     15   0.07000   
12                   JORNADA   object    21695       5     15   0.07000   
4                INSTITUCION   object    21695      16     15   0.07000   
11                 ZONA_SEDE   object    21695       2     15   0.07000   
5                       DANE   object    21695       2     15   0.07000   
10               CONSECUTIVO   object    21695       2     15   0.07000   
9           CODIGO_DANE_SEDE   object    21695       3     15   0.07000   
6                 CALENDARIO   object    21695       1     15   0.07000   
14                     GRUPO   object    21667     178     43   0.20000   
25                 INTERNADO   object    20572       3   1138   5.24000   
21               SISBEN_TRES  float64    20347    2479   1363   6.28000   
28                      SRPA   object    13610       3   8100  37.31000   
27  APOYO_ACADEMICO_ESPECIAL   object    13608       6   8102  37.32000   
30               PAIS_ORIGEN   object     6653       4  15057  69.36000   
16                    MOTIVO   object      958      10  20752  95.59000   
26              NUM_CONTRATO   object        6       2  21704  99.97000   
18                  FECHAFIN  float64        0       0  21710 100.00000   

         mean        min        25%        50%        75%        max      std  
0  2019.00000 2019.00000 2019.00000 2019.00000 2019.00000 2019.00000  0.00000  
2         NaN        NaN        NaN        NaN        NaN        NaN      NaN  
3         NaN        NaN        NaN        NaN        NaN        NaN      NaN  
7         NaN        NaN        NaN        NaN        NaN        NaN      NaN  
22        NaN        NaN        NaN        NaN        NaN        NaN      NaN  
19        NaN        NaN        NaN        NaN        NaN        NaN      NaN  
17        NaN        NaN        NaN        NaN        NaN        NaN      NaN  
1         NaN        NaN        NaN        NaN        NaN        NaN      NaN  
20        NaN        NaN        NaN        NaN        NaN        NaN      NaN  
29        NaN        NaN        NaN        NaN        NaN        NaN      NaN  
24        NaN        NaN        NaN        NaN        NaN        NaN      NaN  
23        NaN        NaN        NaN        NaN        NaN        NaN      NaN  
13    7.54062   -2.00000    3.00000    6.00000    9.00000   26.00000  6.47236  
15        NaN        NaN        NaN        NaN        NaN        NaN      NaN  
8         NaN        NaN        NaN        NaN        NaN        NaN      NaN  
12        NaN        NaN        NaN        NaN        NaN        NaN      NaN  
4         NaN        NaN        NaN        NaN        NaN        NaN      NaN  
11        NaN        NaN        NaN        NaN        NaN        NaN      NaN  
5         NaN        NaN        NaN        NaN        NaN        NaN      NaN  
10        NaN        Na

In [11]:
df_analysis(df_students_2020, key_columns=["NUI"], analysis_type="complete")


Analysis header
--------------------------------------------------------------------------------
- Dataset shape:			 19994 rows and 32 columns
- Total of NaN values:			 80991
- Percentage of NaN:			 12.66 %
- Total of infinite values:		 0
- Total of empty columns:		 1
- Total of empty rows:			 0
- Total of full duplicates rows:	 0
- Unique indexes:			 True
- Memory usage:				 4.9+ MB

- The key(s):	 ['NUI'] is not present multiple times in the dataframe. 
		 It CAN be used as a primary key.


Detailed analysis
--------------------------------------------------------------------------------


name     type  records  unique  # NaN     % NaN  \
0                        ANO    int64    19994       1      0   0.00000   
2                     ESTADO   object    19994       6      0   0.00000   
3                  JERARQUIA   object    19994       3      0   0.00000   
25           FUENTE_RECURSOS   object    19994       6      0   0.00000   
24       MATRICULACONTRATADA   object    19994       6      0   0.00000   
22                    GENERO   object    19994       2      0   0.00000   
7                     SECTOR   object    19994       1      0   0.00000   
20                   ESTRATO   object    19994       8      0   0.00000   
19                       NUI   object    19994   19994      0   0.00000   
30              DISCAPACIDAD   object    19994      13      0   0.00000   
17                  FECHAINI   object    19994   14270      0   0.00000   
1                        ETC   object    19994       1      0   0.00000   
13                 GRADO_COD  float64    19993      19      1   0.01000   
15                    MODELO   object    19992       5      2   0.01000   
14                     GRUPO   object    19992     136      2   0.01000   
12                   JORNADA   object    19992       5      2   0.01000   
11                 ZONA_SEDE   object    19992       2      2   0.01000   
10               CONSECUTIVO   object    19992       2      2   0.01000   
9           CODIGO_DANE_SEDE   object    19992       4      2   0.01000   
8                       SEDE   object    19992      56      2   0.01000   
6                 CALENDARIO   object    19992       1      2   0.01000   
5                       DANE   object    19992       2      2   0.01000   
4                INSTITUCION   object    19992      16      2   0.01000   
21               SISBEN_TRES  float64    19111    2461    883   4.42000   
26                 INTERNADO   object    18797       3   1197   5.99000   
29                      SRPA   object    16654       3   3340  16.71000   
28  APOYO_ACADEMICO_ESPECIAL   object    16649       6   3345  16.73000   
23            TIPO DE SANGRE   object    14451      10   5543  27.72000   
31               PAIS_ORIGEN   object    12635       8   7359  36.81000   
16                    MOTIVO   object      674       4  19320  96.63000   
27              NUM_CONTRATO   object        5       2  19989  99.97000   
18                  FECHAFIN  float64        0       0  19994 100.00000   

         mean        min        25%        50%        75%        max      std  
0  2020.00000 2020.00000 2020.00000 2020.00000 2020.00000 2020.00000  0.00000  
2         NaN        NaN        NaN        NaN        NaN        NaN      NaN  
3         NaN        NaN        NaN        NaN        NaN        NaN      NaN  
25        NaN        NaN        NaN        NaN        NaN        NaN      NaN  
24        NaN        NaN        NaN        NaN        NaN        NaN      NaN  
22        NaN        NaN        NaN        NaN        NaN        NaN      NaN  
7         NaN        NaN        NaN        NaN        NaN        NaN      NaN  
20        NaN        NaN        NaN        NaN        NaN        NaN      NaN  
19        NaN        NaN        NaN        NaN        NaN        NaN      NaN  
30        NaN        NaN        NaN        NaN        NaN        NaN      NaN  
17        NaN        NaN        NaN        NaN        NaN        NaN      NaN  
1         NaN        NaN        NaN        NaN        NaN        NaN      NaN  
13    6.22813   -1.00000    3.00000    6.00000    9.00000   26.00000  4.63803  
15        NaN        NaN        NaN        NaN        NaN        NaN      NaN  
14        NaN        NaN        NaN        NaN        NaN        NaN      NaN  
12        NaN        NaN        NaN        NaN        NaN        NaN      NaN  
11        NaN        NaN        NaN        NaN        NaN        NaN      NaN  
10        NaN        NaN        NaN        NaN        NaN        NaN      NaN  
9         NaN        NaN    

In [12]:
df_analysis(df_students_2021, key_columns=["NUI"], analysis_type="complete")


Analysis header
--------------------------------------------------------------------------------
- Dataset shape:			 21562 rows and 26 columns
- Total of NaN values:			 70984
- Percentage of NaN:			 12.66 %
- Total of infinite values:		 0
- Total of empty columns:		 1
- Total of empty rows:			 0
- Total of full duplicates rows:	 0
- Unique indexes:			 True
- Memory usage:				 4.3+ MB

- The key(s):	 ['NUI'] is not present multiple times in the dataframe. 
		 It CAN be used as a primary key.


Detailed analysis
--------------------------------------------------------------------------------


name     type  records  unique  # NaN     % NaN           mean  \
0            ANO    int64    21562       1      0   0.00000     2021.00000   
1            ETC   object    21562       1      0   0.00000            NaN   
2         ESTADO   object    21562       9      0   0.00000            NaN   
3      JERARQUIA   object    21562       3      0   0.00000            NaN   
23  DISCAPACIDAD   object    21562      16      0   0.00000            NaN   
21        GENERO   object    21562       2      0   0.00000            NaN   
20        PER_ID    int64    21562   21562      0   0.00000 66557297.13167   
7         SECTOR   object    21562       1      0   0.00000            NaN   
17           NUI   object    21562   21562      0   0.00000            NaN   
15      FECHAINI   object    21562    7976      0   0.00000            NaN   
18       ESTRATO   object    21561       7      1   0.00000            NaN   
11     GRADO_COD  float64    21560      20      2   0.01000        7.27055   
13        MODELO   object    21541       8     21   0.10000            NaN   
10       JORNADA   object    21541       5     21   0.10000            NaN   
9      ZONA_SEDE   object    21541       2     21   0.10000            NaN   
8           SEDE   object    21541      55     21   0.10000            NaN   
6     CALENDARIO   object    21541       1     21   0.10000            NaN   
5           DANE   object    21541       2     21   0.10000            NaN   
4    INSTITUCION   object    21541      16     21   0.10000            NaN   
12         GRUPO   object    21540     187     22   0.10000            NaN   
19   SISBEN_TRES  float64    20930    2415    632   2.93000        5.47834   
22          SRPA   object    19401       3   2161  10.02000            NaN   
24   PAIS_ORIGEN   object    17320      12   4242  19.67000            NaN   
14        MOTIVO   object      900       6  20662  95.83000            NaN   
25   Unnamed: 25   object        9       9  21553  99.96000            NaN   
16      FECHAFIN  float64        0       0  21562 100.00000            NaN   

            min            25%            50%            75%            max  \
0    2021.00000     2021.00000     2021.00000     2021.00000     2021.00000   
1           NaN            NaN            NaN            NaN            NaN   
2           NaN            NaN            NaN            NaN            NaN   
3           NaN            NaN            NaN            NaN            NaN   
23          NaN            NaN            NaN            NaN            NaN   
21          NaN            NaN            NaN            NaN            NaN   
20 927709.00000 61149809.00000 68949210.00000 75116811.00000 82761284.00000   
7           NaN            NaN            NaN            NaN            NaN   
17          NaN            NaN            NaN            NaN            NaN   
15          NaN            NaN            NaN            NaN            NaN   
18          NaN            NaN            NaN            NaN            NaN   
11     -2.00000        3.00000        6.00000        9.00000       26.00000   
13          NaN            NaN            NaN            NaN            NaN   
10          NaN            NaN            NaN            NaN            NaN   
9           NaN            NaN            NaN            NaN            NaN   
8           NaN            NaN            NaN            NaN            NaN   
6           NaN            NaN            NaN            NaN            NaN   
5           NaN            NaN            NaN            NaN            NaN   
4           NaN            NaN            NaN            NaN            NaN   
12          NaN            NaN            NaN            NaN            NaN   
19     -1.00000       -1.00000       -1.00000       -1.00000       98.00000   
22          NaN            NaN            NaN            NaN            NaN   
24          NaN            NaN            NaN            NaN            NaN   
14       

In [13]:
df_analysis(df_students_2022, key_columns=["NUI"], analysis_type="complete")


Analysis header
--------------------------------------------------------------------------------
- Dataset shape:			 21111 rows and 34 columns
- Total of NaN values:			 117357
- Percentage of NaN:			 16.35 %
- Total of infinite values:		 0
- Total of empty columns:		 1
- Total of empty rows:			 0
- Total of full duplicates rows:	 0
- Unique indexes:			 True
- Memory usage:				 5.5+ MB

- The key(s):	 ['NUI'] is not present multiple times in the dataframe. 
		 It CAN be used as a primary key.


Detailed analysis
--------------------------------------------------------------------------------


name     type  records  unique  # NaN     % NaN  \
0                        ANO    int64    21111       1      0   0.00000   
7                     SECTOR   object    21111       1      0   0.00000   
31              DISCAPACIDAD   object    21111      15      0   0.00000   
23                    GENERO   object    21111       2      0   0.00000   
22                    PER_ID    int64    21111   21111      0   0.00000   
19                       NUI   object    21111   21111      0   0.00000   
1                        ETC   object    21111       1      0   0.00000   
17                  FECHAINI   object    21111    7746      0   0.00000   
2                     ESTADO   object    21111       8      0   0.00000   
3                  JERARQUIA   object    21111       3      0   0.00000   
13                 GRADO_COD  float64    21109      19      2   0.01000   
6                 CALENDARIO   object    21101       1     10   0.05000   
4                INSTITUCION   object    21101      16     10   0.05000   
9           CODIGO_DANE_SEDE   object    21101       4     10   0.05000   
5                       DANE   object    21101       2     10   0.05000   
8                       SEDE   object    21101      54     10   0.05000   
15                    MODELO   object    21101       7     10   0.05000   
12                   JORNADA   object    21101       5     10   0.05000   
11                 ZONA_SEDE   object    21101       2     10   0.05000   
10               CONSECUTIVO   object    21101       2     10   0.05000   
20                   ESTRATO   object    21100       8     11   0.05000   
25       MATRICULACONTRATADA   object    21093       6     18   0.09000   
26           FUENTE_RECURSOS   object    21093       5     18   0.09000   
14                     GRUPO   object    21077     178     34   0.16000   
30                      SRPA   object    19403       3   1708   8.09000   
29  APOYO_ACADEMICO_ESPECIAL   object    19393       6   1718   8.14000   
27                 INTERNADO   object    18974       3   2137  10.12000   
32               PAIS_ORIGEN   object    17700      15   3411  16.16000   
24            TIPO DE SANGRE   object    15244      13   5867  27.79000   
21                 SISBEN IV   object     2472      38  18639  88.29000   
16                    MOTIVO   object      724       4  20387  96.57000   
33               Unnamed: 33   object        9       9  21102  99.96000   
28              NUM_CONTRATO   object        7       2  21104  99.97000   
18                  FECHAFIN  float64        0       0  21111 100.00000   

             mean          min            25%            50%            75%  \
0      2022.00000   2022.00000     2022.00000     2022.00000     2022.00000   
7             NaN          NaN            NaN            NaN            NaN   
31            NaN          NaN            NaN            NaN            NaN   
23            NaN          NaN            NaN            NaN            NaN   
22 69293050.55767 927709.00000 64055607.50000 70688322.00000 76899696.50000   
19            NaN          NaN            NaN            NaN            NaN   
1             NaN          NaN            NaN            NaN            NaN   
17            NaN          NaN            NaN            NaN            NaN   
2             NaN          NaN            NaN            NaN            NaN   
3             NaN          NaN            NaN            NaN            NaN   
13        7.05017     -1.00000        3.00000        6.00000        9.00000   
6             NaN          NaN            NaN            NaN            NaN   
4             NaN          NaN            NaN            NaN            NaN   
9             NaN          NaN            NaN            NaN            NaN   
5             NaN          NaN            NaN            NaN            NaN   
8             NaN          NaN            NaN            NaN            NaN   
15            NaN          NaN            NaN          

<div class="alert alert-block alert-success">
    <p><b>Observations</b></p>
    <p>Below, some points observed:</p>
    <p>
        <ul style="list-style-type: square;">
            <li>it appears, the number of children differs in each list. It is necessary to validate how much students are into each dataset.</li>
            <li>The amount of variables do not match between datasetsets</li>
            <li>There is a primary key into all dataset ['NUI']  that is not present multiple times in each dataframe.</li>
        </ul>
    </p>
    <p>Now, a little detail for each of the datasets</p>
    <p>
        <ul style="list-style-type: square;">
            <li><b>df_alumnos_2019</b></li>
            <ul style="list-style-type: disc;">
               <li>Dataset shape: 21710 rows and 31 columns</li>
               <li>Missing values: 98130 / 14.58%</li>
               <li>Key(s) column(s): Nui can be used as a primary key</li>
            </ul>
        </ul>
        <ul style="list-style-type: square;">
            <li><b>df_alumnos_2020</b></li>
            <ul style="list-style-type: disc;">
               <li>Dataset shape: 19994 rows and 32 columns</li>
               <li>Missing values: 80991 / 12.66 %</li>
               <li>Key(s) column(s): Nui can be used as a primary key</li>
            </ul>
        </ul>
        <ul style="list-style-type: square;">
            <li><b>df_alumnos_2021</b></li>
            <ul style="list-style-type: disc;">
               <li>Dataset shape: 21562 rows and 26 columns</li>
               <li>Missing values: 70984 / 12.66 %</li>
               <li>Key(s) column(s): Nui can be used as a primary key</li>
            </ul>
        </ul>
        <ul style="list-style-type: square;">
            <li><b>df_alumnos_2021</b></li>
            <ul style="list-style-type: disc;">
               <li>Dataset shape: 21111 rows and 34 columns</li>
               <li>Missing values: 117357 / 16.35 %</li>
               <li>Key(s) column(s): Nui can be used as a primary key</li>
            </ul>
        </ul>
    </p>
</div>

## 3. Cleaning